hax/hare_cov/hare_coverage.py

```diff
@@ -0,0 +1,354 @@
+    }
+
+    def __init__(self, cov_report_type: str = "xml",
+                 report_dir: str = "hare_coverage_report",
+                 st_map: Dict[str, str] = source_test_map_hare.copy()):
+        """
+        Sets coverage report type, directory where report will generated and
+        unit test directories source modules.
+        :param cov_report_type: coverage type xml or html
+        :param report_dir: directory where report will generated
+        :param st_map: source module to test module map
+        """
+        self.cov_report_type: str = cov_report_type
+        self.src_tst_map: Dict[str, str] = st_map
+        self.report_dir = report_dir
+        self.report = self.report_dir + sep
+        self.report += (
+            "html_report" if self.cov_report_type == "html"
+            else "coverage.xml")
+        self.utils = Utility()
+        logging.info("Running Python Coverage Tool.")
+
+    def cleanup(self) -> None:
+        """
+        Cleaning up old coverage reports.
+        :return: None
+        """
+        logging.info("Cleaning up the report.")
+        try:
+            if self.utils.is_dir_exist(self.report_dir) is True:
+                self.utils.run_cmd(f"rm -rf {self.report_dir}")
+        except Exception as e:
+            logging.error(f"Error {e} while cleanup.")
+
+    def run_ut(self) -> None:
+        """
+        Execute unit testcases.
+        :return: None
+        """
+        logging.info("Running UT's.")
+        dir_scanner = DirectoryScanner("test")
+        for test_dir_path in dir_scanner.search_dir(getcwd()):
+            self.utils.run_cmd(f"pytest {test_dir_path}")
+
+    def run_coverage(self) -> None:
+        """
+        Running coverage tool over UTs.
+        :return: None
+        """
+        logging.info("Running Coverage.")
+        command_line: str
+        for index, (src, tst) in enumerate(self.src_tst_map.items()):
+            if index == 0:
+                command_line = f"pytest --cov={src} " \
+                                    f"--cov-report={self.cov_report_type}:" \
+                                    f"{self.report} {tst}"
+            else:
+                command_line = f"pytest --cov={src} " \
+                                    f"--cov-report={self.cov_report_type}:" \
+                                    f"{self.report} --cov-append {tst}"
+            self.utils.run_cmd(command_line)
+
+    def check_report(self) -> None:
+        """
+        Validate the report generated.
+        :return: None
+        """
+        logging.info("Checking report.")
+        if self.utils.is_dir_exist(self.report) is True:
+            logging.info("Python Code Coverage Report generated successfully "
+                         f"at {self.report!a}.")
+            print("Python Code Coverage Report generated successfully "
+                  f"at {self.report!a}.")
+        else:
+            raise FailedToGenerateReport("Failed to generate the Coverage"

```

Comment:
Will the user understand the reason why it failed?

Reply:
We hare coverage.log file where error logs are kept. Mostly script is executing linux os commands through run_cmd. We are raising exception if process returns error return code.

Reply:
Then I'd suggest to add at least a note where to find the details on this error into the exception text. `Failed to generate the Coverage` doesn't help anybody - neither programmers who will be called for support, nor the end users will understand what exactly went wrong and what can be the cause

Reply:
Added more print where to look for the error logs.